packages/client-utils/src/vstorage-client.js

```diff
@@ -0,0 +1,489 @@
+        height,
+        kind,
+        path,
+        rpcAddress,
+      );
+
+    return codec.response.decode(decodeBase64(result.response.value));
+  };
+
+  return { getLatestHeight, queryAbci };
+};
+
+/**
+ * @template {any} T
+ * @param {Array<T>} values
+ */
+const makeRoundRobin = values => {
+  let currentIndex = 0;
+
+  return () => {
+    const element = values[currentIndex];
+    currentIndex = (currentIndex + 1) % values.length;
+    return element;
+  };
+};
+
+/**
+ * @template {any} T
+ * @param {{queryClient: ReturnType<typeof makeQueryClient>}} powers
+ * @param {string} path
+ * @param {Partial<{ compat: boolean }>} [options]
+ * @returns {StreamTopic<T>}
+ */
+const makeStreamTopic = ({ queryClient }, path, options) => {
+  /**
+   * @param {Height} blockHeight
+   * @param {StreamCell['values']} values
+   */
+  function* allValuesFromCell(blockHeight, values) {
+    for (const value of values.reverse())
+      yield /** @type {Update<T>} */ ({
+        blockHeight,
+        value: JSON.parse(value),
+      });
+  }
+
+  /**
+   * @param {boolean} compatMode
+   * @param {ReturnType<typeof parseValue>} value
+   */
+  const assertValueStructure = (compatMode, value) => {
+    if (!(isStreamCell(value) || compatMode))
+      throw Error(`Expected a stream cell, got ${quote(value)}`);
+  };
+
+  /**
+   * @param {Height} [minimum]
+   * @param {Height} [maximum]
+   */
+  async function* iterate(minimum, maximum) {
+    /** @type {Array<Update<T>>} */
+    let values = [];
+
+    for await (const value of reverseIterate(maximum, minimum, true))
+      values = [value, ...values];
+
+    for (const value of values) {
+      assertValueStructure(!!options?.compat, {
+        blockHeight: String(value.blockHeight),
+        values: [/** @type string */ (value.value)],
+      });
+      yield value;
+    }
+  }
+
+  /**
+   * @param {Height} [maximum]
+   * @param {Height} [minimum]
+   * @param {boolean} [compatMode]
+   */
+  async function* reverseIterate(
+    maximum,
+    minimum,
+    compatMode = options?.compat,
+  ) {
+    let blockHeight = maximum;
+    /** @type {Height} */
+    let currentBlockHeight;
+    /** @type {string | undefined} */
+    let lastValue;
+
+    await null;
+
+    if (!minimum || minimum < MINIMUM_HEIGHT) minimum = MINIMUM_HEIGHT;
+
+    if (maximum && maximum < minimum)
+      throw createError(
+        `${INVALID_HEIGHT_ERROR_MESSAGE} ${maximum}`,
+        blockHeight || (await queryClient.getLatestHeight()),
+        PATHS.DATA,
+        path,
+      );
+
+    while (true) {
+      /** @type {string | undefined} */
+      let value;
+
+      try {
+        ({ value } = await queryClient.queryAbci(path, {
+          height: blockHeight,
+          kind: PATHS.DATA,
+        }));
+      } catch (err) {
+        console.error(err);
+        break;
+      }
+
+      if (!value) break;
+
+      let parsedValue = parseValue(value);
+      assertValueStructure(!!compatMode, parsedValue);
```

Comment:
If you are concerned with invalid input for `compatMode`, consider coercing to boolean once up front by renaming the option `compatModeOption` and:
```js
const compatMode = !!compatModeOption;
```

That’ll capture the result and avoid synchronously reinvoking protocols like `valueOf` if the clever user passes an object. But, in my opinion, this is sufficiently protected by TypeScript and the `!!` is superfluous, whereas the manual coercion is still vulnerable to reëntrance so falls short of what we get from interface guards and patterns when the caller might be an attacker.